src/tojson.jl

```diff
@@ -4,53 +4,117 @@ using JSON
 """
     tojson(output::String, ts::ReportingTestSet)
 
-Takes user output and a ReportingTestSet and converts it to a JSON string as expected by the interface.
+Takes user output and a ReportingTestSet and converts it to a JSON string as
+expected by the interface.
+
+Here's a brief summary of the schema we're following:
+
+{
+    version: 2
+    status: pass | fail | error
+    message?: "a descriptive message emitted only if none of the tests ran (e.g. a syntax error in the file)"
+    tests: [
+        {
+            name: "a human-readable name"
+            status: pass | fail | error
+            message?: "Describes the result of the test if it did not pass"
+            output?: "the stdout for this test"
+            test_code?: "snippet of code that failed that student can try locally"
+                (required for failing or erroring concept exercise tests)
+            task_id?: "a concept exercise thing we don't support yet"
+                (bump version to 3 if we support this)
+        },
+    ]
+}
+
+Note:
+    We currently populate test_code only with the failing or erroring `@test`
+    expression. This means that we don't give all the necessary information if
+    the test relies on definitions outside of that expression, e.g.
+
+    ```julia
+    @testset begin
+        robot = Robot()
+        @test robot.rename!()
+    end
+    ```
+
+    For that to make sense, the student would really need the entirety of the
+    enclosing testset, but we will only give them the string
+    `@test robot.rename!()`
+
+For more information, check the reference:
+
+https://github.com/exercism/docs/blob/main/building/tooling/test-runners/interface.md
 """
 function tojson(output::String, ts::ReportingTestSet)
-    tests = Dict{String, Union{String, Nothing}}[]
-    for s in ts.results
-        status = "pass"
-        message = nothing
-        for t in s.results
-            if typeof(t) == Test.Fail
+    if length(ts.results) == 1 && ts.results[1] isa Test.Error
+        # There has been a syntax error or similar and no tests have run.
+        # Otherwise ts.results[1] will be a ReportingTestSet.
+        return JSON.json(Dict(
+            "version" => 2,
+            "status" => "error",
+            "message" => ts.results[1].backtrace,
+            "tests" => [],
+        ), 4)
+    end
```

Comment:
The indent here is wrong, the rest of the file is indent by an extra level.

Reply:
I think you're just misreading this? `walk!()` etc. are inner functions of `tojson`.

Reply:
Oh ok, I haven't seen docstrings for inner functions before. Then it's fine.

Reply:
It's a bit unusual, but I think it's a good idea :shrug: